### ISSUE TITLE
[10.0e] TaxCloud module assigns wrong tax on SO confirm

### ISSUE THREAD

**Issue by matt454357** [odoo/odoo#18084]
**Impacted versions:** 10.0e (TaxCloud module)

**Steps to reproduce:**
1. Create sales order
2. Click Confirm Sale

**Current behavior:**
Sales tax changes from 9.25% to 9.2485714286%

**Expected behavior:**
Sales tax should not change

**Video/Screenshot link (optional):**
https://www.youtube.com/watch?v=TxtDzvwi2Hs

**Hint:**
The problem is in this file:
https://github.com/odoo/enterprise/blob/10.0/sale_account_taxcloud/models/sale_order.py

Try changing precision_digits from 4 to 2...
```python
if float_compare(line.tax_id.amount, tax_rate, precision_digits=4):
```
changes to...
```python
if float_compare(line.tax_id.amount, tax_rate, precision_digits=2):
```

And, this will also need to change:
```python
'name': 'Tax %s %%' % (tax_rate),
```

**Comment by Yenthe666** [comment]
@tivisse I believe this is for you?

**Comment by tivisse** [comment]
Hello @matt454357 

I think that

**Comment by taxcloud** [comment]
**Question:** how could this happen, unless the implementation is attempting to reverse engineer a tax rate from a TaxCloud TaxAmount response.

**Please confirm:** is the Odoo implementation calculation ating a tax rate based upon TaxCloud API TaxAmout responses?

**Why is this important?**
Odoo machinery should NOT reverse engineer tax rates (because several states have odd rounding rules). More importantly, reverse engineering a tax rate suggests a "blanket" application of this derived rate, regardless of product taxability.

In many  states, applicable tax rates vary based upon the category of goods, and even extended price, of items being sold.

If an Odoo implementation is _deriving_ tax rates from a generic Lookup with an arbitrary sales price than most TaxCloud compliance capabilities would be irreparably compromised (including indemnification).

Thanks!

-David

**Comment by tivisse** [comment]
Hi @taxcloud 

We don't try to reverse engineer the tax rates from the previous Taxcloud responses. We just convert the returned amount into a tax rate on the related sales order (after having deducted the discounted price, etc). So there is no magic in there. On the Sales Order validation and on the Invoice validation, a call to Taxcloud is made to validate the taxes for all the Sales Order lines and the returned amounts are converted to a rate on our record.

**Comment by tivisse** [comment]
Hello @matt454357 

Sorry for the delay, your issue has been fixed in 10.0 at https://github.com/odoo/enterprise/commit/681c6c1edb872fa9593ff7fb82c2e4a8ae7af141

Have a nice day,

Yannick.

**Comment by matt454357** [comment]
Hello @tivisse
Please reopen this issue.  It still is not working properly, as shown in the following screenshots.  Before confirming, the tax rate is 6.88% for both lines.  Confirming changes the tax rates to 6.87% and 6.85%.  The real tax rate in Chester NJ is 6.875%.

![draft](https://user-images.githubusercontent.com/7604932/31238049-d953a2a6-a9b5-11e7-8078-054ab9d72459.png)

![confirmed](https://user-images.githubusercontent.com/7604932/31238057-e1f91832-a9b5-11e7-8211-f58431c6f222.png)

**Comment by taxcloud** [comment]
Looks like the only issue is display rounding. But cannot explain the variety of rounding results (6.88, 6.87, 6.85) unless it is averaging across multiple factors.

**Comment by matt454357** [comment]
Upon further review of the code, I believe we should revert your fix.
Rounding the tax rate to the 2nd decimal place could introduce a significant error on the total tax amount.  This will guarantee perfect accuracy on the tax amount.  We need to use full float precision.

The final tax rate is based on the rounded amount from TaxCloud.  Many sales tax rates in the US have 4 significant figures (e.g. 6.875%).  So, the calculated rate will only be accurate on prices greater than 1000.

Now, my customers will complain if I show some wierd tax rate on the sales order.  The easiest fix for this problem is to hide the tax rate on sales order, and only show the final total tax amount.

**Comment by bazemoreb** [comment]
Hi @matt454357  - yes, definitely agree this is still broken.  I had made some notes on this in a different thread also i'm looking for.  It looks to me like on the validation step, the code is looking to reverse calculate the taxes as a means of checking.  the rounding comes from that.

**Comment by bazemoreb** [comment]
Here was my testing/example:
Create a partner with the address of 1521 S Burr St, Mitchell, SD 57301 (tax rate should be 6.5% - confirmed on taxcloud's website). Now create a product with price of $25 and a product with price of $19.95. Add those to either a quotation or an invoice (it seems to follow same steps and same result). You will see that a tax is created of 6.5%. Now confirm quote or validate invoice. 


What is the current behavior that you observe? 
You will notice that initially the tax is 6.5%, but after a confirm/validate, this changes to 6.52% (see video). I've figured out why I think. 

My example uses the two item lines I referenced above: 

Voip seats - 25.00 
Plantronics adapter – 19.95 

If you multiply by 6.5% you get: 

25*.065 = 1.625 
19.95*0.65 = 1.29675 

If you then round that to two digits you get 1.63 and 1.30 respectively. 

Then you divide those BACK in: 

1.63/25 = 6.52% 
1.3/19.95 = 6.516% (rounds to 6.52%) 

So it looks like on validate/confirm, the module is doing some sort of double check by dividing to see if the tax rates match. If they do NOT match it is creating a NEW tax rate.

**Comment by matt454357** [comment]
Your calculations are correct.  The tax rate is obviously wrong, but it still gives an accurate total because the amount error is in fractions of pennies (which get dropped by the decimal precision setting).  The effect of the python code is to use the **AMOUNT** from TaxCloud, and so the **RATE** is not important.

There's an easy fix: don't show the rate in the sales order report.  This assumes you don't have a good reason to require the rate to show on the sales order.  So, we can ignore the rate, and know that the amount is correct.

**Comment by matt454357** [comment]
The reason the rate changes from quotation to confirmed sales order:

Quotations estimate the tax rate based on a $100 purchase.  That pushes the rounding error beyond the 2nd decimal place, where Odoo rounds the tax name (e.g. 6.88% for a rate of 6.875%).

When confirming the sales order, the code get's the actual tax amount from TaxCloud.  If the price of an individual item is less than $100, then we can't set a precise rate.  It doesn't really matter, as explained above, but it is ugly.

p.s. Odoo would be smart to keep tax rates at the 3rd decimal, so tax estimates on quotations for items over $1,000 would be more accurate.  Maybe in version 12?

**Comment by bazemoreb** [comment]
Hmm - would have to look through code, but I'm not sure why it would make sense to get the tax amount and use an arbitrary number to calculate the tax percentage.  I would think that taxcloud would be returning an actual tax percentage (also - as you say, this should be three decimal rounding, as this is common).  My assumption was we pass the address to taxcloud, and are returned a decimal representing the tax amount.

**Comment by matt454357** [comment]
TaxCloud returns a tax **amount**, not a tax **percentage rate**.  That's explained by David Campbell, of TaxCloud, in his post above.  Unfortunately, the entire Odoo system was designed to use percentage rates.

TaxCloud is making some very complex computations (rounding rules, tax holidays, extended price considerations, regressive rates, product categories, etc.).  A simple percentage rate wouldn't handle the complexity.

In other words, Odoo SA had to bend their percentage based system out of shape to make it work with the complexities of US sales tax.

Disclaimer: I'm just another user, trying to figure out how to make this module work.  I don't speak for Odoo or TaxCloud.

**Comment by tivisse** [comment]
Thanks to all for the different feedbacks and replies. 

We're working on this issue. It's mainly coming from the fact that we work with a tax amount (already rounded) and we work in odoo with tax percentages. I personally think that the proposal by @matt454357 is legit. I still have to discuss this with @antonylesuisse about it.

The other issue (different tax rates before and after confirmation) is due to a bad design of the module. The tax rates are stored on the fiscal position to retrieve the value without making a call to taxcloud is there's a match with the following parameters (State, Zip, TIC category). As the tax computation takes into account more parameters that the 3 I've enumerate, I also think that we should get rid of this 'magic guessing' and only compute taxes on orders/invoices validation on the /shop/payment/ 

Thanks for your patience.

Yannick.

**Comment by tivisse** [comment]
Hello @taxcloud 

In fact, I think that it could be possible to solve our issue here if it was possible to retrieve the taxed amount from the call to 'Lookup' without rounding. This way we do not create a lot of taxes in the database (like 6.5 and 6.52) due to rounding errors from the inverse computation.

Should it be possible to have a mode raw to get the taxed amount without rounding ?

**Comment by taxcloud** [comment]
Hey all.

- First, if storing a rate, it should stored to 5 decimal places, not 3 (Click around [our find a rate map, particularly in New Mexico](https://taxcloud.net/#Find-a-Rate)).
- Second, we stopped returning TaxAmount with fractions (decimals) of pennies last year because some developers were truncating after 2 decimal places. I would generally agree that the rate is not necessary for display, though I respect some may be used to seeing it.
- Finally, we have an experimental endpoint (part of our upcoming 2.0 release) that supports secondary transaction details (returning tax amount, rate, and applicable tax jurisdictions, on a per CartItem basis). May be easiest to comment-out the rate display until 2.0 launches. For planning purposes, it is also important to note our next release will also include non-US jurisdictions.

Thank you!
-David

**Comment by bazemoreb** [comment]
Hi All - looks like I should have read the whole thread before replying above, but it all makes sense now.  Glad to hear that 2.0 will have the ability to get the actual tax rate as that seems like a good option for sure.

WRT the showing of the amount on the sales orders/invoices etc., maybe makes sense to leverage the "description" field and just set the name to the default (we use "Sales Tax" for example) and that way all invoices and such will just show "Sales Tax" with no rate, but when looking as an employee you can see.  Just a thought!

**Comment by tivisse** [comment]
Thanks for the reply @taxcloud Could you already estimate a release date ? That 2.0 version would be very welcome :)

@bazemoreb I've also thought about something like this, but there's an unique constraint on a tax name (Which I can understand if we want to avoid to taxes with the same name and different rates).

I'm gonna clear the situation with @fpodoo and @antonylesuisse and come back to you with a solution. 

Once again thanks for participating to this thread.

**Comment by tivisse** [comment]
@taxcloud Could you confirm that we're speaking about the same precision.

In Odoo, a tax rate is encoded like 3.568, we suppose that it is in percent. (instead of storing 0.03568)

**Comment by tivisse** [comment]
The issue should now be solved with these fixes:

1/ Store tax rate amount with 3 digits instead of 2
https://github.com/odoo/enterprise/commit/a1c808df853ee7ed5ac56dd57cb7004dff2e2c7c

2/ Stop magic guessing tax rates before SO/inv confirmation
https://github.com/odoo/enterprise/commit/e0a97bdc44fb88c767e4ef53ac55ddfd9775a1a5

This will avoid the tax rate to pass from 6.5 to 6.52% on the confirmation.

3/ Allow to compute the taxes on a sales order
 https://github.com/odoo/enterprise/commit/24b58117a5901680bb33ca6349012a728484de1b

This was already existing on the invoice to update the taxes without confirming it. However, this change require the module account_taxcloud to be update to see the changes.

4/ Display 'Sales Tax' on the SO/Inv report instead of 'Tax 6.52%'
https://github.com/odoo/enterprise/commit/07545da6ac8d31a4bc35e32031ec98f9d5509df1

We can't do much more in a stable release. The customers will see 'Sales Tax' on their printed reports (And the taxed amount is correct). The employees will see the applied tax rate on the line.

Any feedback is welcomed.

Have a nice day,

Yannick.

**Comment by bazemoreb** [comment]
Thanks @tivisse - Will look through this and do some testing when possible.  Also not sure if you saw the PR I did to fix the tax generation from subscriptions.  First PR I've ever done, and looks like it has some conflicts but this is pretty important for anyone using subscriptions and accountcloud.  Easy fix, though.

https://github.com/odoo/enterprise/pull/1473

**Comment by taxcloud** [comment]
This is great news!

Following up on the question regarding decimal precision, the states expect accuracy to 5 decimals for the rate (not the dollars).

So, the example of 3.568% may be generally acceptable, but may not be accurate, as the true rate could be 3.56761%... and then you have States with rounding rules that do not match ordinary bank rounding (rounding up at 9 rather than 5).

This is why TaxCloud responds with dollars and cents of tax amounts due, to shield developers from all such intricacies.

And as a final warning to displaying tax rates, some states have partial rate reductions for items under or over specific dollar amounts, meaning the effective tax rate could be 6.5% for a specific dollar amount portion of the sales, and a reduced 1.23456% for the remaining portion. Doing math on these situations will provoke customer service tickets because the effective tax rate, will look like it is incorrect for the average consumer.

Let us be your tax geeks. It’s all we do, and our service is free (by default).

Thank you for all of your efforts, and please let us know if we can do anything to help.

-David

**Comment by tivisse** [comment]
Hi @bazemoreb 

Sorry for the late reply, I didn't see your message before today. I'm gonna check this and give you a feedback on the PR you've created.

Hi @taxcloud 

Thanks for the reply. Finally this tax precision thing is much more complicated than we thought. The important point here is that the taxed amount is always correct :).

Have a nice day.

**Comment by vcloud9ray** [comment]
> some states have partial rate reductions for items under or over specific dollar amounts, meaning the effective tax rate could be 6.5% for a specific dollar amount portion of the sales, and a reduced 1.23456% for the remaining portion.

Can you share an example of this - I have a Customer who wants to research this particular

**Comment by bazemoreb** [comment]
Hi Ray - Yes, so that exact scenario is why @taxcloud is saying they don't want to send over the percentage, and why odoo is changing now so that they don't save the percentage (as i understand).  What taxcloud does is send back the amount of the tax which accounts for your example.

**Comment by vcloud9ray** [comment]
@bazemoreb - understood. I have Customer that is fielding questions from the people who pay the Invoices like "I calculated the percentage by diving the tax into the total and it was X% and it should be Y%". I am hoping that being able to point to an example where a state doesn't charge a flat rate will help them educate people why that won't always work.

**Comment by bazemoreb** [comment]
Makes sense!  I don't have a concrete example unfortunately, but I do know the TN sales tax is a bad one.  Here is an article on the tiered sales tax I found.  I'm sure the taxcloud guys can get something better for ya, though.  https://www.tn.gov/revenue/article/sales-and-use-tax-single-article

**Comment by vcloud9ray** [comment]
@taxcloud - sorry to ping you here, but do you have a strategy for correctly requesting Sales Tax for a Customer Deposit - the downpayment (may or not be refundable) that some companies require for projects, services or high price / new product creation?   See https://github.com/odoo/odoo/issues/20986

**Comment by taxcloud** [comment]
Thanks for the ping - we just replied to [#20986](https://github.com/odoo/odoo/issues/20986).

The answer: **TIC 10005 (Gift Card)**

> Gift certificates are treated like cash. No sales tax is collected at the time the certificate is purchased. When a customer presents a gift certificate, the customer receives a credit against the total cost of the purchase and pays the difference, if any. The sales tax applies to the total amount of the purchase, not the amount due after the certificate is deducted.

**Comment by taxcloud** [comment]
Replying to the earlier request from @vcloud9ray, we can cite many examples, but here are a few easy ones:

1. In the Commonwealth of Massachusetts, clothing items are tax-exempt below $175 (per item), however, every penny above $175 is taxed at 6.25%. So, a $200 sweater would have a tax amount of $1.56 (which would divide out to an effective tax rate of .0078%).

2. In the state of New York, shipping charges are taxable, unless **everything** being purchased is exempt. So if you sell $10,000 of t-shirts, and charge $1,000 for shipping, the entire $11,000 transaction is exempt. However, if you add even a single taxable item, say a $2 sticker, then your $11,002 transaction would be subjected to $88.93 of tax (shipping and the sticker attracting the NYC 8.875% sales tax rate, however dividing that tax amount by the entire $11,002 transaction would suggest a tax rate of .008%).

And then, once you get all of these rules in place, some city decide to enact a sales tax holiday for a particular class of goods, but the state doesn't adopt the same sales tax holiday.

Our _raison d'etre_ is that developers and retailers shouldn't have to track this stuff. The states that make up these crazy rules should work with us to verify and certify that we have implemented their crazy laws correctly. States benefit by knowing the tax compliance is done properly, while developers and retailers benefit with accurate tax amounts as well as audit relief and indemnification from the States (in the unlikely event we make an error).

**Comment by tivisse** [comment]
Hi @taxcloud 

I hope you're doing well.

Do you have any news about the 'v2.0' release ? 

We have more customers complaining about the fact that Taxcloud returns the taxed amount with 2 decimals instead of 3 or doesn't return the tax rates.

Looking forward to hearing from you,

Yannick.

**Comment by jx2y** [comment]
@tivisse @taxcloud, did this thread die or move? We have an implementation of Odoo11 online and are experiencing the same issues as everything above, showing the wrong sales tax due to the rounding factor involved. Any insight on when a fix will be implemented for the online version of Odoo? Thank you!! 

Jason

**Comment by tivisse** [comment]
Hi @jx2y 

Still no new about the release Taxcloud 2.0. We don't have any other choice than deal with it. (Means having small rounding factor errors on the tax rates BUT having the correct taxed amount).

Thanks for your understanding.

**Comment by matt454357** [comment]
I customized the invoice report to hide the rate and show only the tax amount.  Let the customer do their own math. No complaints yet.

**Comment by ray-odoo** [comment]
This is how Odoo is now handling it with the 2019 changes in v11 and v12 and master.

**Comment by bazemoreb** [comment]
Yep we rebuilt ours to do the same. :/
